design/design-doc-flexible-claims-and-composition.md

```diff
@@ -0,0 +1,945 @@
+
+The below example uses the infrastructure operator defined `MySQLInstance` claim
+from the [above example] along with a `CompositeResourceClass`. The use of a
+`CompositeResourceClass` rather than a `ResourceClass` signals that multiple
+primitive managed resources must be provisioned to satisfy the resource claim.
+
+```yaml
+---
+apiVersion: database.example.org/v1alpha1
+kind: MySQLInstance
+metadata:
+  namespace: default
+  name: sql
+  annotations:
+    crossplane.io/external-name: very-private-database
+spec:
+  engineVersion: "5.7"
+  storageGB: 10
+  region: us-west
+  writeConnectionSecretToRef:
+    name: sql
+  classRef:
+    apiVersion: infrastructure.crossplane.io/v1alpha1
+    kind: CompositeResourceClass
+    name: private-mysql-server
+---
+apiVersion: infrastructure.crossplane.io/v1alpha1
+kind: CompositeResourceClass
+metadata:
+  name: private-mysql-server
+spec:
+  # Each CompositeResourceClass defines how resources are dynamically
+  # provisioned for exactly one kind of resource claim; there is a one-to-many
+  # claim-kind-to-composite-resource-class-instance relationship.
+  claim:
+    apiVersion: database.example.org/v1alpha1
+    kind: MySQLInstance
+  # The array of primitive managed resources that will be dynamically
+  # provisioned to satisfy MySQLInstance claims that use this class. Like the
+  # above ResourceClass, each entry is an object containing a 'base' and an
+  # optional 'patch'.
+  resources:
+  - base:
+      apiVersion: azure.crossplane.io/v1alpha3
+      kind: ResourceGroup
+      metadata:
+      spec:
+        location: West US
+        providerRef:
+          name: example
+        reclaimPolicy: Delete
+    patch:
+    - fromClaimFieldPath: "spec.region"
+      toResourceFieldPath: "spec.forProvider.location"
+      transforms:
+      - type: map
+        map:
+          us-west: "West US"
+          us-east: "East US"
+  - base:
+      apiVersion: database.azure.crossplane.io/v1beta1
+      kind: MySQLServer
+      spec:
+        forProvider:
+          administratorLogin: myadmin
+          # A resource selector allows a resource dynamically provisioned using
+          # a particular composite resource class to reference other resources
+          # provisioned by the same class. In this case MySQLServer resources
+          # provisioned using this class will select the ResourceGroup
+          # provisioned by the same class and claim.
+          resourceGroupNameSelector:
+            matchComposite: true
+          location: West US
+          sslEnforcement: Disabled
+          version: "5.6"
+          sku:
+            tier: Basic
+            capacity: 1
+            family: Gen5
+          storageProfile:
+            storageMB: 20480
+        writeConnectionSecretToRef:
+          namespace: crossplane-system
+        providerRef:
+          name: example
+        reclaimPolicy: Delete
+    patch:
+    - fromClaimFieldPath: ".metadata.uid"
```

Comment:
I think `fieldPath` doesn't prefix it via `.` like `JSONPath`

Reply:
Good call - will fix.